File: DemoApp/DemoApp/NativeAssetsDetailsViewController.m
//
//  NativeAssetsDetailsViewController.m
//  DemoApp
//
//  Created by Glispa GmbH on 27/03/2017.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

#import "NativeAssetsDetailsViewController.h"
#import "NativeAssetsPlusView.h"
#import "AppControllerManagement.h"


@interface NativeAssetsDetailsViewController ()

@property (weak, nonatomic) IBOutlet UIView* adContainerView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *nativeContainerHeightConstraint;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *nativeContainerTopConstraint;

@property (weak, nonatomic) UIView *nativeView;
@end

@implementation NativeAssetsDetailsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}


- (void)loadAds {
    __typeof__(self) __weak wSelf = self;
    [AvocarrotSDK.sharedSDK loadNativeAdWithAdUnitId:self.adUnitId
                                             success:^UIView * _Nonnull(AVONativeAssets *
                                                                        _Nonnull nativeAssets) {
                __typeof__(self) sSelf = wSelf;

                NativeAssetsPlusView *nativeView = [sSelf getNativeAdContainerForNativeAd:nativeAssets];
                NSArray *clickableSubviews = [nativeView clickableSubViews];

                [sSelf.nativeView removeFromSuperview];
                sSelf.nativeView = nativeView;
                                                 
                [sSelf renderAd];

                [[[[nativeAssets onImpression:^{
                    NSLog(@"Native Assets! Impressed");
                }] onClick:^{
                    NSLog(@"Native Assets! Clicked");
                }] onLeftApplication:^{
                    NSLog(@"Native Assets! Left application");
                }] registerViewForInteraction:nativeView
                         forClickableSubviews:clickableSubviews];

                [nativeAssets registerAdChoicesView:nativeView.privacyInfoContainerView];

                return nativeView;
            }
            failure:^(AVOError *_Nonnull error) {
                NSLog(@"Native Assets loading error: %@", [error localizedDescription]);
            }];
}

#pragma mark - Private

- (NativeAssetsPlusView *)getNativeAdContainerForNativeAd:(AVONativeAssets *)ad {

    Class adViewClass = [NativeAssetsPlusView class];
    NSString *className = NSStringFromClass(adViewClass);

    NativeAssetsPlusView *nativeAdContainer = (NativeAssetsPlusView *) [[NSBundle mainBundle] loadNibNamed:className owner:nil options:nil][0];
    [nativeAdContainer fillWithNativeAd:ad];

    return nativeAdContainer;
}


- (void)renderAd {
    if (![self.nativeView superview]) {
        self.nativeContainerHeightConstraint.constant = 300;
        [self.adContainerView layoutIfNeeded];

        self.nativeView.frame = self.adContainerView.bounds;
        self.nativeView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [self.adContainerView addSubview:self.nativeView];
        [self.adContainerView layoutIfNeeded];

        [self showBannerWithAnimation:YES];
    }
}


- (void)showBannerWithAnimation:(BOOL)animated {
    self.nativeContainerTopConstraint.constant = -self.nativeContainerHeightConstraint.constant;
    if (animated) {
        [UIView animateWithDuration:0.2 animations:^{
            [self.view layoutIfNeeded];
        }];
    } else {
        [self.view layoutIfNeeded];
    }
}

@end


@interface NativeAssetsDetailsViewController (AppManagement) <AppControllerManagement>
@end

@implementation NativeAssetsDetailsViewController (AppManagement)

- (void)applyUserData:(NSDictionary <NSString *, NSObject *> *)userData {
    if ([userData[@"adUnitId"] isKindOfClass:[NSString class]]) {
        self.adUnitId = (NSString *)userData[@"adUnitId"];
    }
    
    [self loadAds];
}

@end

File: DemoApp/DemoAppSwift/NativeAssetsDetailsViewController.swift
//
//  NativeAssetsDetailsViewController.swift
//  DemoApp
//
//  Created by Glispa GmbH on 27/03/2017.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

import UIKit
import AvocarrotNativeAssets

class NativeAssetsDetailsViewController: UIViewController {

    @IBOutlet weak var adContainerView: UIView!

    @IBOutlet weak var nativeContainerHeightConstraint: NSLayoutConstraint!
    @IBOutlet weak var nativeContainerTopConstraint: NSLayoutConstraint!

    public var adUnitId: String!
    fileprivate var nativeView: UIView?

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    func loadAds() {
        nativeView?.removeFromSuperview()

        AvocarrotSDK.shared.loadNativeAd(withAdUnitId: adUnitId, success: { [weak self] (
                nativeAd) in
            guard let sSelf = self,
                  let nativeView = sSelf.getNativeAdContainer(forNativeAd: nativeAd) else {
                return nil
            }

            let clickableSubviews = nativeView.clickableSubViews()

            sSelf.nativeView?.removeFromSuperview()
            sSelf.nativeView = nativeView

            sSelf.renderBannerAd()

            nativeAd.onClick({
                print("Native Assets! Clicked")
            }).onImpression({
                print("Native Assets! Impressed")
            }).onLeftApplication({
                print("Native Assets! Left application")
            }).registerView(forInteraction: nativeView, forClickableSubviews: clickableSubviews)
              .registerAdChoicesView(nativeView.privacyInfoContainerView)

            return nativeView
        }, failure: { error in
            print("Native Assets error: \(error.localizedDescription)")
        })
    }

    // MARK: - Private

    fileprivate func getNativeAdContainer(forNativeAd ad: AVONativeAssets?) -> NativeAssetsPlusView? {
        let nativeAdContainer = NativeAssetsPlusView.loadFromNib()
        nativeAdContainer?.fill(withNativeAd: ad)
        return nativeAdContainer
    }

    fileprivate func renderBannerAd() {
        if nativeView?.superview == nil {
            nativeContainerHeightConstraint.constant = 300
            adContainerView.layoutIfNeeded()

            nativeView!.frame = self.adContainerView.bounds
            nativeView!.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            adContainerView.addSubview(self.nativeView!)
            adContainerView.layoutIfNeeded()

            showBannerWithAnimation(true)
        }
    }

    fileprivate func showBannerWithAnimation(_ animated: Bool) {
        nativeContainerTopConstraint.constant = -self.nativeContainerHeightConstraint.constant
        if animated {
            UIView.animate(withDuration: 0.2, animations: {
                self.view.layoutIfNeeded()
            })
        } else {
            self.view.layoutIfNeeded()
        }
    }

}

extension NativeAssetsDetailsViewController: AppControllerManagement {
    func applyUserData(_ userData: [String:Any]) {
        if let bannerId = userData["adUnitId"] as? String {
            adUnitId = bannerId
        }

        loadAds()
    }
}
